### PR TITLE
planner: fix bug when unfolding wildcard in view definiton

### LINKED ISSUE #24933: view returns unexpected error when containing count(*)
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` SQL
drop table if exists t1;
CREATE TABLE `t1` (
  `a` int(11) DEFAULT NULL,
  `b` varchar(32) DEFAULT NULL
);
create view v3 as select * from (select a, count(*) from t1 where b >= 'c' group by a) qn;
select * from v3;
```

### 2. What did you expect to see? (Required)

No error

### 3. What did you see instead (Required)
```
mysql> select * from v3;
ERROR 1356 (HY000): View 'test.v3' references invalid table(s) or column(s) or function(s) or definer/invoker of view lack rights to use them
```
### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by guo-shaoge**
Signed-off-by: guo-shaoge <shaoge1994@163.com>

### What problem does this PR solve?

Issue Number: close #24933 

Problem Summary: When wildcard and subquery(e.g. select some_expr/agg/win_func from xxx) is used in view definition. We got unknown column error.

For example:

```
create view v as select * from (select count(*) from t1) s;
select * from v; // got error
create view v as select * from (select c1 or 0 from t1) s;
select * from v; // got error
```

**Why**: check the result of show create view. The `*` is unfolded to be `s.count(*)`, but in the subquery, the select field is `count(1)`. The column name is not same. (**PS**: This problem exists for other agg func/window func ...)

    SELECT `s`.`count(*)` AS `count(*)` FROM (SELECT COUNT(1) AS `count(*)` FROM `test`.`t1`) AS `s`

### What is changed and how it works?

What's Changed: **Add alias name for select filed when create view**. So the view definition will be:

    SELECT `s`.`count(*)` FROM (SELECT COUNT(1) AS `count(*)` FROM `test`.`t1`) AS `s`

How it works:
1. After `buildProjection()` in `buildSelect()`, we add alias name for each select field. The new added function assumes there is a **one-to-one mapping** between select fields and proj.OutputNames().

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects: None

### Release note 

- planner: fix bug when unfolding wildcard in view definiton

### CODE REVIEW

**Comment by guo-shaoge** [comment]
@lzmhhh123 PTAL~

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -1032,6 +1032,14 @@ func (b *PlanBuilder) buildProjectionFieldNameFromExpressions(ctx context.Contex
 		// When the query is select t.a from t group by a; The Column Name should be a but not t.a;
 		return agg.Args[0].(*ast.ColumnNameExpr).Name.Name, nil
 	}
+	if _, ok := field.Expr.(*ast.AggregateFuncExpr); ok {
```
Comment (winoros) [inline planner/core/logical_plan_builder.go:1035]: Please add some comments for this part

**Review by winoros — CHANGES_REQUESTED** [review]

**Review by lzmhhh123 — APPROVED** [review]

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -1027,10 +1027,11 @@ func (b *PlanBuilder) buildProjectionFieldNameFromColumns(origField *ast.SelectF
 }
 
 // buildProjectionFieldNameFromExpressions builds the field name when field expression is a normal expression.
-func (b *PlanBuilder) buildProjectionFieldNameFromExpressions(ctx context.Context, field *ast.SelectField) (model.CIStr, error) {
+func (b *PlanBuilder) buildProjectionFieldNameFromExpressions(ctx context.Context, field *ast.SelectField) (model.CIStr, bool, error) {
```
Comment (winoros) [inline planner/core/logical_plan_builder.go:1030]: Comment whats the newly added return value mean

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -3124,6 +3132,10 @@ func unfoldWildStar(field *ast.SelectField, outputName types.NameSlice, column [
 				}}
 			colName.SetType(col.GetType())
 			field := &ast.SelectField{Expr: colName}
+			// Alias name is only to ensure the view definition is correct.
+			if capFlag&canExpandAST != 0 && name.SelectField != nil {
+				name.SelectField.AsName = colName.Name.Name
```
Comment (winoros) [inline planner/core/logical_plan_builder.go:3137]: We are passing the `SelectField` as a pointer. Is this in-place modification safe?

Comment (guo-shaoge) [reply planner/core/logical_plan_builder.go:3137]: Here I want to change the `ast.SelectField` by adding an alias name. e.g. from `count(1)` to `count(1) AS count(*)`. So We have to use pointer make it take effects.

I don't very understand the specific meaning of safety.(like thread safety or some other scenes?)

**Review by qw4990 — APPROVED** [review]

**Pull request merged**